src/Simulation/Simulators/OpenSystemsSimulator/Decompositions/IntrinsicInterface.qs

```diff
@@ -0,0 +1,210 @@
+                    within {
+                        H(target);
+                    } apply {
+                        S(target);
+                    }
+                } apply {
+                    Controlled Z(controls, target);
+                }
+            } else {
+                ApplyWithLessControlsA(Controlled Y, (controls, target));
+            }
+        }
+    }
+
+    operation Z(target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            Native.Z(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                Z(target);
+            } elif Length(controls) == 1 {
+                within {
+                    H(target);
+                } apply {
+                    Native.CNOT(controls[0], target);
+                }
+            } elif Length(controls) == 1 {
+                // [Page 15 of arXiv:1206.0758v3](https://arxiv.org/pdf/1206.0758v3.pdf#page=15)
+                Adjoint T(controls[0]);
+                Adjoint T(controls[1]);
+                Controlled X([target], controls[0]);
+                T(controls[0]);
+                Controlled X([controls[1]], target);
+                Controlled X([controls[1]], controls[0]);
+                T(target);
+                Adjoint T(controls[0]);
+                Controlled X([controls[1]], target);
+                Controlled X([target], controls[0]);
+                Adjoint T(target);
+                T(controls[0]);
+                Controlled X([controls[1]], controls[0]);
+            } else {
+                ApplyWithLessControlsA(Controlled Z, (controls, target));
+            } 
+        }
+    }
+
+    operation S(target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            Native.S(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                S(target);
+            } elif (Length(controls) == 1) {
+                T(controls[0]);
+                T(target);
+                within {
+                    Controlled X([controls[0]], target);
+                } apply {
+                    Adjoint T(target);
+                }
+            } else {
+                ApplyWithLessControlsA(Controlled S, (controls, target));
+            }
+        }
+    }
+
+    operation T(target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            Native.T(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                T(target);
+            } else {
+                // TODO: Decompositions of `Controlled T` currently used in Q#
+                //       target packages rely on R1Frac, which is not yet
+                //       implemented in experimental simulation.
+                fail "Controlled T operations are not yet supported.";
+            }
+        }
+    }
+
+    // NB: We separate out this operation to avoid hardware targeting rewrite
```

Comment:
Might be good to include a mention of https://github.com/microsoft/qsharp-compiler/issues/768 here.

Reply:
Done.